README.md
# subsystem_entropy_epsplots
An .eps plot program for the subsystem_entropy dataset. Just copy the folder into the job folder and let the magic happen.


changelog.txt
### 18.09.2017
- added short time inlay to entropy plot
- added fit to entropy plot (and removed it again afterwards)

### 15.09.2017
- added plot_all.sh for easy plotting of all the stuff
- added en_eps.py for energy related plots

### 05.09.2017
- added entropy eps plot and averaging
- added default.ini to the files list so older datasets can be analyzed as well


pyplot_eps/en_eps.py
import numpy as np
import os
from mpEntropy import mpSystem
import matplotlib as mpl
from matplotlib.pyplot import cm
import matplotlib.pyplot as plt
from scipy.signal import savgol_filter
# This is a workaround until scipy fixes the issue
import warnings

warnings.filterwarnings(action="ignore", module="scipy", message="^internal gelsd")

# Load sysVar
sysVar = mpSystem("../interact_0.ini", plotOnly=True)

# Create plot folder
pltfolder = "./epsplots/"
if not os.path.exists(pltfolder):
    os.mkdir(pltfolder)

print("Plotting", end='')
mpl.use('Agg')
# styles and stuff
avgstyle = 'dashed'
avgsize = 0.6
expectstyle = 'solid'
expectsize = 1

loavgpercent = sysVar.plotLoAvgPerc  # percentage of time evolution to start averaging
loavgind = int(loavgpercent * sysVar.dataPoints)  # index to start at when calculating average and stddev
loavgtime = np.round(loavgpercent * (sysVar.deltaT * sysVar.steps * sysVar.plotTimeScale), 2)

if sysVar.boolPlotAverages:
    print(' with averaging from Jt=%.2f' % loavgtime, end='')
fwidth = sysVar.plotSavgolFrame
ford = sysVar.plotSavgolOrder
params = {
    'legend.fontsize': sysVar.plotLegendSize,
    'font.size': sysVar.plotFontSize,
    'mathtext.default': 'rm'  # see http://matplotlib.org/users/customizing.html
}
plt.rcParams['agg.path.chunksize'] = 0
plt.rcParams.update(params)
plt.rc('text', usetex=True)
plt.rc('font', **{'family': 'sans-serif', 'sans-serif': ['Arial']})

bool_eigenvalues = False  # plot eigenvalues and decomposition?
if os.path.isfile('../data/hamiltonian_eigvals.txt'):
    bool_eigenvalues = True
    energy_array = np.loadtxt('../data/hamiltonian_eigvals.txt')

bool_total = False  # plot total energy?
if os.path.isfile('../data/total_energy.txt'):
    bool_total = True
    total_energy_array = np.loadtxt('../data/total_energy.txt')
elif os.path.isfile('../data/energy.txt'):  # old file path
    bool_total = True
    total_energy_array = np.loadtxt('../data/energy.txt')

# eigenvalues and spectral decomposition
if bool_eigenvalues:
    fig, ax1 = plt.subplots()
    energy_markersize = 0.7
    energy_barsize = 0.06
    if sysVar.dim != 1:
        energy_markersize *= (2.0 / np.log10(sysVar.dim))
        energy_barsize *= (4.0 / np.log10(sysVar.dim))
    ax1.plot(energy_array[:, 0], energy_array[:, 1], linestyle='none', marker='o', ms=energy_markersize, color='blue')
    ax1.set_ylabel(r'Energy / J')
    ax1.set_xlabel(r'Eigenvalue index $n$')
    plt.grid(False)
    ax1.set_xlim(xmin=-(len(energy_array[:, 0]) * (5.0 / 100)))
    if np.shape(energy_array)[0] > 2:
        ax2 = ax1.twinx()
        ax2.bar(energy_array[:, 0], energy_array[:, 2], alpha=0.8, color='red', width=energy_barsize, align='center')
        ax2.set_ylabel(r'$|c_n|^2$')
        # inlay with small region around maximum
        ax_inlay = plt.axes([0.62, 0.6, 0.28, 0.28])
        index_range = np.floor(sysVar.dim / 200)
        index_maximum = np.amax(energy_array[:, 2])
        index_lo = index_maximum - index_range
        index_hi = index_maximum + index_range
        decomp_max = np.max(energy_array[index_lo:index_hi, 2])
        decomp_min = np.min(energy_array[index_lo:index_hi, 2])
        ax_inlay.bar(energy_array[index_lo:index_hi, 0], energy_array[index_lo:index_hi, 2], color='red', width=energy_barsize * 10, align='center')
        ax_inlay.set_xticks([energy_array[index_lo, 0], energy_array[index_hi, 0]])
        ax_inlay.set_yticks([decomp_min,decomp_max])
    plt.tight_layout()
    ###
    plt.savefig(pltfolder + 'energy_eigenvalues.eps', format='eps', dpi=1000)
    plt.clf()
    print('.', end='', flush=True)

    # Eigenvalue decomposition with energy x-axis
    plt.bar(energy_array[:, 1], energy_array[:, 2], alpha=0.8, color='red', width=energy_barsize, align='center')
    plt.xlabel(r'Energy / J')
    plt.ylabel(r'$|c_E|^2$')
    plt.grid(False)
    plt.xlim(xmin=-(np.abs(energy_array[0, 1] - energy_array[-1, 1]) * (5.0 / 100)))
    plt.tight_layout()
    ###
    plt.savefig(pltfolder + 'energy_decomposition.eps', format='eps', dpi=1000)
    plt.clf()
    print('.', end='', flush=True)

# Total energy
if bool_total:
    en0 = total_energy_array[0, 1]
    total_energy_array[:, 1] -= en0
    en0_magnitude = np.floor(np.log10(np.abs(en0)))
    en0 /= np.power(10, en0_magnitude)
    magnitude = np.floor(np.log10(np.max(np.abs(total_energy_array[:, 1]))))
    plt.plot(total_energy_array[:, 0] * sysVar.plotTimeScale, total_energy_array[:, 1] / (np.power(10, magnitude)),
             linewidth=0.6)
    plt.figtext(0.9, 0.85, r'$E_0 / J = %.2f \cdot 10^{%i}$' % (en0, en0_magnitude), horizontalalignment='right',
                verticalalignment='bottom')
    plt.ylabel(r'$E_{tot} - E_0 / (J \cdot 10^{%i})$' % magnitude)
    plt.xlabel(r'$J\,t$')
    plt.grid(False)
    plt.tight_layout()
    ###
    plt.savefig(pltfolder + 'energy_total.eps', format='eps', dpi=1000)
    plt.clf()
    print('.', end='', flush=True)

print(" done!")


pyplot_eps/ent_eps.py
import numpy as np
import os
from mpEntropy import mpSystem
import matplotlib as mpl
from matplotlib.pyplot import cm
import matplotlib.pyplot as plt
import matplotlib.mlab as mlab
from scipy.stats import norm
from scipy.signal import savgol_filter
from scipy.optimize import curve_fit
# This is a workaround until scipy fixes the issue
import warnings

warnings.filterwarnings(action="ignore", module="scipy", message="^internal gelsd")

# Load sysVar
sysVar = mpSystem("../interact_0.ini", plotOnly=True)

# Create plot folder
pltfolder = "./epsplots/"
if not os.path.exists(pltfolder):
    os.mkdir(pltfolder)

print("Plotting", end='')
mpl.use('Agg')
# styles and stuff
avgstyle = 'dashed'
avgsize = 0.6
expectstyle = 'solid'
expectsize = 1

loavgpercent = sysVar.plotLoAvgPerc  # percentage of time evolution to start averaging
loavgind = int(loavgpercent * sysVar.dataPoints)  # index to start at when calculating average and stddev
loavgtime = np.round(loavgpercent * (sysVar.deltaT * sysVar.steps * sysVar.plotTimeScale), 2)

if sysVar.boolPlotAverages:
    print(' with averaging from Jt=%.2f' % loavgtime, end='')
fwidth = sysVar.plotSavgolFrame
ford = sysVar.plotSavgolOrder
params = {
    'legend.fontsize': sysVar.plotLegendSize,
    'font.size': sysVar.plotFontSize,
    'mathtext.default': 'rm'  # see http://matplotlib.org/users/customizing.html
}
plt.rcParams['agg.path.chunksize'] = 0
plt.rcParams.update(params)
plt.rc('text', usetex=True)
plt.rc('font', **{'family': 'sans-serif', 'sans-serif': ['Arial']})

ent_array = np.loadtxt('../data/entropy.txt')
# multiply step array with time scale
step_array = ent_array[:, 0] * sysVar.plotTimeScale

#### Complete system Entropy
if os.path.isfile('../data/total_entropy.txt'):
    totent_array = np.loadtxt('../data/total_entropy.txt')
    plt.plot(totent_array[:, 0] * sysVar.plotTimeScale, totent_array[:, 1] * 1e13, linewidth=0.6, color='r')

    plt.grid()
    plt.xlabel(r'$J\,t$')
    plt.ylabel(r'Total system entropy $/ 10^{-13}$')
    plt.tight_layout()
    ###
    plt.savefig(pltfolder + 'entropy_total.eps', format='eps', dpi=1000)
    plt.clf()
    print('.', end='', flush=True)

### Subsystem Entropy
fldat = open(pltfolder + 'ent_fluctuation_N' + str(sysVar.N) + '.txt', 'w')
fldat.write('N_tot: %i\n' % sysVar.N)
avg = np.mean(ent_array[loavgind:, 1], dtype=np.float64)
stddev = np.std(ent_array[loavgind:, 1], dtype=np.float64)
fldat.write('ssent_average: %.16e\n' % avg)
fldat.write('ssent_stddev: %.16e\n' % stddev)
fldat.write('ssent_rel._fluctuation: %.16e\n' % (stddev / avg))
fldat.close()

plt.plot(step_array[:], ent_array[:, 1], linewidth=0.3, color='r')
plt.grid()
if sysVar.boolPlotAverages:
    tavg = savgol_filter(ent_array[:, 1], fwidth, ford)
    plt.plot(step_array[loavgind:], tavg[loavgind:], linewidth=avgsize, linestyle=avgstyle, color='black')
plt.xlabel(r'$J\,t$')
plt.ylabel('Subsystem entropy')
plt.tight_layout()
plt.savefig(pltfolder + 'entropy_subsystem.eps', format='eps', dpi=1000)
plt.clf()


# Subsystem entropy with inlay
max_time = 1.5
max_ind = int(max_time / step_array[-1] * len(step_array))

plt.plot(step_array[:], ent_array[:, 1], linewidth=0.3, color='r')
if sysVar.boolPlotAverages:
    tavg = savgol_filter(ent_array[:, 1], fwidth, ford)
    plt.plot(step_array[loavgind:], tavg[loavgind:], linewidth=avgsize, linestyle=avgstyle, color='black')
plt.xlabel(r'$J\,t$')
plt.ylabel('Subsystem entropy')
a = plt.axes([.5, .3, .4, .4])
plt.plot(step_array[:max_ind], ent_array[:max_ind, 1], linewidth=0.3, color='r')
plt.savefig(pltfolder + 'entropy_subsystem_inlay.eps', format='eps', dpi=1000)
plt.clf()
print('.', end='', flush=True)

# histogram of fluctuations
n, bins, patches = plt.hist(ent_array[loavgind:, 1] - avg, 51, normed=1, rwidth=0.8, align='mid')
(mu, sigma) = norm.fit(ent_array[loavgind:, 1] - avg)
y = mlab.normpdf(bins, mu, sigma)
l = plt.plot(bins, y, 'r--', linewidth=2)
mu_magnitude = np.floor(np.log10(np.abs(mu)))
mu /= np.power(10, mu_magnitude)
sigma_magnitude = np.floor(np.log10(sigma))
sigma /= np.power(10, sigma_magnitude)
plt.figtext(0.92, 0.82,
            '$\mu = %.2f \cdot 10^{%i}$\n$\sigma = %.2f \cdot 10^{%i}$' % (mu, mu_magnitude, sigma, sigma_magnitude),
            ha='right', va='bottom', multialignment="left", bbox=dict(fc="none"))
plt.xlabel(r'$S_{sub}$ fluctuation amplitude')
plt.ylabel(r'Probability density')
plt.tight_layout()
plt.savefig(pltfolder + 'entropy_subsystem_fluctuations.eps', format='eps', dpi=1000)
plt.clf()
print('.', end='', flush=True)

print(" done!")


pyplot_eps/occ_eps.py
import numpy as np
import os
from mpEntropy import mpSystem
import matplotlib as mpl
from matplotlib.pyplot import cm
import matplotlib.pyplot as plt
from scipy.signal import savgol_filter
# This is a workaround until scipy fixes the issue
import warnings

warnings.filterwarnings(action="ignore", module="scipy", message="^internal gelsd")

# Load sysVar
sysVar = mpSystem("../interact_0.ini", plotOnly=True)

# Create plot folder
pltfolder = "./epsplots/"
if not os.path.exists(pltfolder):
    os.mkdir(pltfolder)

print("Plotting", end='')
mpl.use('Agg')
# styles and stuff
avgstyle = 'dashed'
avgsize = 0.6
expectstyle = 'solid'
expectsize = 1

loavgpercent = sysVar.plotLoAvgPerc  # percentage of time evolution to start averaging
loavgind = int(loavgpercent * sysVar.dataPoints)  # index to start at when calculating average and stddev
loavgtime = np.round(loavgpercent * (sysVar.deltaT * sysVar.steps * sysVar.plotTimeScale), 2)

if sysVar.boolPlotAverages:
    print(' with averaging from Jt=%.2f' % loavgtime, end='')
fwidth = sysVar.plotSavgolFrame
ford = sysVar.plotSavgolOrder
params = {
    'legend.fontsize': sysVar.plotLegendSize,
    'font.size': sysVar.plotFontSize,
    'mathtext.default': 'rm'  # see http://matplotlib.org/users/customizing.html
}
plt.rcParams['agg.path.chunksize'] = 0
plt.rcParams.update(params)
plt.rc('text', usetex=True)
plt.rc('font', **{'family': 'sans-serif', 'sans-serif': ['Arial']})

occ_array = np.loadtxt('../data/occupation.txt')
# multiply step array with time scale
step_array = occ_array[:, 0] * sysVar.plotTimeScale

### Single-level occupation numbers

for i in range(0, sysVar.m):
    plt.plot(step_array, occ_array[:, i + 1], label=r'$n_' + str(i) + '$', linewidth=0.5)
    if sysVar.boolPlotAverages:
        tavg = savgol_filter(occ_array[:, i + 1], fwidth, ford)
        plt.plot(step_array[loavgind:], tavg[loavgind:], linewidth=avgsize, linestyle=avgstyle, color='black')

plt.ylabel(r'Occupation number')
plt.xlabel(r'$J\,t$')
plt.legend(loc='upper right')
plt.grid()
plt.tight_layout()
plt.savefig(pltfolder + 'occupation_single.eps', format='eps', dpi=1000)
plt.clf()
print('.', end='', flush=True)

### Traced out (bath) occupation numbers
for i in sysVar.kRed:
    plt.plot(step_array, occ_array[:, i + 1], label=r'$n_' + str(i) + '$', linewidth=0.6)
    if sysVar.boolPlotAverages:
        tavg = savgol_filter(occ_array[:, i + 1], fwidth, ford)
        plt.plot(step_array[loavgind:], tavg[loavgind:], linewidth=avgsize, linestyle=avgstyle, color='black')

plt.ylabel(r'Occupation number')
plt.xlabel(r'$J\,t$')
plt.legend(loc='lower right')
plt.grid()
plt.tight_layout()
plt.savefig(pltfolder + 'occupation_bath.eps', format='eps', dpi=1000)
plt.clf()
print('.', end='', flush=True)

### Leftover (system) occupation numbers

for i in np.arange(sysVar.m)[sysVar.mask]:
    plt.plot(step_array, occ_array[:, i + 1], label=r'$n_' + str(i) + '$', linewidth=0.6)
    if sysVar.boolPlotAverages:
        tavg = savgol_filter(occ_array[:, i + 1], fwidth, ford)
        plt.plot(step_array[loavgind:], tavg[loavgind:], linewidth=avgsize, linestyle=avgstyle, color='black')

plt.ylabel(r'Occupation number')
plt.xlabel(r'$J\,t$')
plt.legend(loc='lower right')
plt.grid()
plt.tight_layout()
plt.savefig(pltfolder + 'occupation_system.eps', format='eps', dpi=1000)
plt.clf()
print('.', end='', flush=True)

### Subsystems occupation numbers
# store fluctuations in a data
fldat = open(pltfolder + 'occ_fluctuation_N'+str(sysVar.N)+'.txt', 'w')
fldat.write('N_tot: %i\n' % sysVar.N)
tmp = np.zeros(len(step_array))
for i in sysVar.kRed:
    tmp += occ_array[:, i + 1]
plt.plot(step_array, tmp, label="bath", linewidth=0.8, color='magenta')

if sysVar.boolPlotAverages:
    tavg = savgol_filter(tmp, fwidth, ford)
    plt.plot(step_array[loavgind:], tavg[loavgind:], linewidth=avgsize, linestyle=avgstyle, color='black')

avg = np.mean(tmp[loavgind:], dtype=np.float64)
stddev = np.std(tmp[loavgind:], dtype=np.float64)
fldat.write('bath_average: %.16e\n' % avg)
fldat.write('bath_stddev: %.16e\n' % stddev)
fldat.write('bath_rel._fluctuation: %.16e\n' % (stddev / avg))

tmp.fill(0)
for i in np.arange(sysVar.m)[sysVar.mask]:
    tmp += occ_array[:, i + 1]
plt.plot(step_array, tmp, label="system", linewidth=0.8, color='darkgreen')

if sysVar.boolPlotAverages:
    tavg = savgol_filter(tmp, fwidth, ford)
    plt.plot(step_array[loavgind:], tavg[loavgind:], linewidth=avgsize, linestyle=avgstyle, color='black')

avg = np.mean(tmp[loavgind:], dtype=np.float64)
stddev = np.std(tmp[loavgind:], dtype=np.float64)
fldat.write('system_average: %.16e\n' % avg)
fldat.write('system_stddev: %.16e\n' % stddev)
fldat.write('system_rel._fluctuation: %.16e\n' % (stddev / avg))

for i in range(sysVar.m):
    avg = np.mean(occ_array[loavgind:, i + 1], dtype=np.float64)
    stddev = np.std(occ_array[loavgind:, i + 1], dtype=np.float64)
    fldat.write('n%i_average: %.16e\n' % (i, avg))
    fldat.write('n%i_stddev: %.16e\n' % (i, stddev))
    fldat.write('n%i_rel._fluctuation: %.16e\n' % (i, (stddev / avg)))
fldat.close()

plt.ylabel(r'Occupation number')
plt.xlabel(r'$J\,t$')
plt.legend(loc='upper right')
plt.grid()
plt.tight_layout()
plt.savefig(pltfolder + 'occupation_systems.eps', format='eps', dpi=1000)
plt.clf()
print('.', end='', flush=True)

### occupation number in levels against level index

occavg = np.loadtxt(pltfolder + 'occ_fluctuation_N'+str(sysVar.N)+'.txt', usecols=(1,))
plt.xlim(-0.1, sysVar.m - 0.9)
for l in range(0, sysVar.m):
    plt.errorbar(l, occavg[int(7 + 3 * l)] / sysVar.N, xerr=None, yerr=occavg[int(8 + 3 * l)] / sysVar.N,
                 marker='o', color=cm.Set1(0))
plt.ylabel(r'Relative level occupation')
plt.xlabel(r'Level index')
plt.grid()
plt.tight_layout()
plt.savefig(pltfolder + 'occupation_distribution.eps', format='eps', dpi=1000)
plt.clf()

plt.xlim(np.min(np.arange(sysVar.m)[sysVar.mask]) - 0.1, np.max(np.arange(sysVar.m)[sysVar.mask]) + 0.1)
for l in np.arange(sysVar.m)[sysVar.mask]:
    plt.errorbar(l, occavg[int(7 + 3 * l)] / sysVar.N, xerr=None, yerr=occavg[int(8 + 3 * l)] / sysVar.N,
                 marker='o', color=cm.Set1(0))
plt.ylabel(r'Relative level occupation')
plt.xlabel(r'Level index')
plt.grid()
plt.tight_layout()
plt.savefig(pltfolder + 'occupation_distribution_sys.eps', format='eps', dpi=1000)
plt.clf()

plt.xlim(np.min(sysVar.kRed) - 0.1, np.max(sysVar.kRed) + 0.1)
for l in sysVar.kRed:
    plt.errorbar(l, occavg[int(7 + 3 * l)] / sysVar.N, xerr=None, yerr=occavg[int(8 + 3 * l)] / sysVar.N,
                 marker='o', color=cm.Set1(0))
plt.ylabel(r'Relative level occupation')
plt.xlabel(r'Level index')
plt.grid()
plt.tight_layout()
plt.savefig(pltfolder + 'occupation_distribution_bath.eps', format='eps', dpi=1000)
plt.clf()

print(" done!")
